### PR TITLE
[sui-sdk] jsonrpc error response classification and handling

### PR DESCRIPTION
**Pull request by wlmyng**
## Description 

Users of our rust sdk should not need to import the jsonrpsee crate just to error match on jsonrpc errors, nor should clients be encouraged to do string matching on the error message.
1. Introduce json_rpc_error::Error, and a conversion from jsonrpsee::Error to json_rpc_error::Error, so we don't make any breaking changes
2. json_rpc_error::Error has some basic api to determine error type for further manual handling
3. It seems like the jsonrpsee crate makes liberal use of error codes in the -32000 to -32099 range, and unfortunately our -32001 transient error code conflicts with their unknown error code. Bumped our transient error code to -32050.

Further improvements:
From here, we can consider providing more detailed classification of errors, if useful. For example, if we receive TRANSACTION_EXECUTION_CLIENT_ERROR_CODE, it's not clear whether this is due to InvalidUserSignature, ObjectsDoubleUsed, or NonRecoverableTransactionError. We can consider populating this in the data field.

## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [x] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes
You can now determine whether an rpc error response is a client or transient error by converting the sui_sdk::error::Error::RpcError into an sui_sdk::error::JsonRpcError. After converting, you can inspect the code, message, and data of the error, and use the methods is_call_error, is_client_error, is_execution_error, and is_transient_error. You can find an example of this usage at sui-sdk/examples/json_rpc_errors.rs. Note that transient errors now map to code -32050, and -32001 is reserved for any unknown errors returned by the rpc.

### CODE REVIEW

File: crates/sui-sdk/src/error.rs [hunk 1]
```diff
@@ -12,13 +13,13 @@ pub enum Error {
...
     #[error(transparent)]
+    JsonRpcError(JsonRpcError),
+    #[error(transparent)]
     BcsSerialisationError(#[from] bcs::Error),
     #[error(transparent)]
     UserInputError(#[from] UserInputError),
     #[error("Subscription error : {0}")]
     Subscription(String),
-    #[error("Encountered error when confirming tx status for {0:?}, err: {1:?}")]
-    TransactionConfirmationError(TransactionDigest, jsonrpsee::core::Error),
```
Comment (amnn) [inline crates/sui-sdk/src/error.rs:21]: Is this removal intended for this PR? It doesn't seem to fit with the rest of the change.

File: crates/sui-sdk/src/json_rpc_error.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use jsonrpsee::types::{error::UNKNOWN_ERROR_CODE, ErrorObjectOwned};
```
Comment (amnn) [inline crates/sui-sdk/src/json_rpc_error.rs:3]: Doesn't this line imply that users of the SDK still need to depend on the jsonrpsee crate to interpret errors?  It should be trivial to get rid of the reference to `UNKNOWN_ERROR_CODE` but not so sure about `ErrorObjectOwned`.

File: crates/sui-sdk/src/json_rpc_error.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    }
+}
+
+impl Error {
+    pub fn is_call_error(&self) -> bool {
+        self.code != UNKNOWN_ERROR_CODE
+    }
+
+    pub fn is_client_error(&self) -> bool {
+        use jsonrpsee::types::error::{
```
Comment (amnn) [inline crates/sui-sdk/src/json_rpc_error.rs:27]: Ditto -- this is a dependency on the crate you're trying to get people to not have to depend on right?

**Review by amnn — COMMENTED** [review]
I may be misunderstanding, but after this PR the `sui-sdk` crate still depends on `jsonrpsee` mostly by referring to its constants.

**Comment by wlmyng** [comment]
> I may be misunderstanding, but after this PR the `sui-sdk` crate still depends on `jsonrpsee` mostly by referring to its constants.

Ah right let me clarify. The sui-sdk will still depend on jsonrpsee, to decouple, I think we'd need to modify the api module to not use the jsonrpsee clients defined on sui-json-rpc so we can control the result returned. This might be worth looking into, had not considered removing the jsonrpsee dependency from the crate tbh  

What I meant in this PR is that for a user of the sdk, to inspect an error, they would need to `use jsonrpsee` in their code, which is not ideal. Instead we wrap the conversions and checks needed to inspect the rpc error in json_rpc_error::Error for clients to use

File: crates/sui-sdk/src/error.rs (hunk 1 above)
Comment (amnn) [reply crates/sui-sdk/src/error.rs:21]: Ah I now see the connection with this PR of removing this transaction, but two questions:

- What about `RpcError`?
- Are we losing structure/information by replacing `TransactionConfirmationError` (a kind of JSON RPC error) with a more generic message?

**Review by amnn — APPROVED** [review]
Ah got it, thanks for clarifying!  Before you land, can you update the release notes as this will be a breaking change for Rust SDK users.  Modulo the remaining question about `TransactionConfirmationError`, this looks good, thanks for the clean-up @wlmyng !

File: crates/sui-sdk/examples/json_rpc_errors.rs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        .get_coins(active_address, coin_type.clone(), None, Some(5))
+        .await;
+    let error = coins.unwrap_err();
+    if let Error::RpcError(rpc_error) = error {
+        let converted: JsonRpcError = rpc_error.into();
+        println!(" *** RpcError ***");
+        println!("{converted}");
+        println!("{}", converted.is_client_error());
+    } else {
+        panic!("Expected Error::RpcError, got {:?}", error);
```
Comment (stefan-mysten) [inline crates/sui-sdk/examples/json_rpc_errors.rs:23]: Might be better to use `bail!("Expected Error::RpcError, got {:?}", error)` instead of panic.

File: crates/sui-sdk/src/error.rs (hunk 1 above)
Comment (wlmyng) [reply crates/sui-sdk/src/error.rs:21]: I'm removing TransactionConfirmationError because it doesn't show up in use anywhere. 

RpcError will still exist. I think eventually we'll want to retire this, especially with the Graphql server coming up, so RpcError -> JsonRpcError, GraphqlError

File: crates/sui-sdk/examples/json_rpc_errors.rs (hunk 4 above)
Comment (wlmyng) [reply crates/sui-sdk/examples/json_rpc_errors.rs:23]: ok, will modify

Comment (amnn) [reply crates/sui-sdk/examples/json_rpc_errors.rs]: Actually... in an example (or test code) it's better to use `panic!` rather than `bail!` because `bail!` will not help you get a stacktrace.  I realise that I'm imparting these rules about `Result` vs `panic!` piecemeal, koan style, so here's the full set of "Ashok's mental model of when to panic":


#### Services

**Never panic** (mostly).  The one exception to this rule is during initialisation (i.e. if you can't bring up the service fully, just panic, don't try and recover and then "half serve" requests).  This means all failures must be recorded as `Result<T, E>`-s where `E` is a **structured error type**, that can be translated to an error response. The `thiserror` crate is good for defining these error types.

#### Libraries

Also heavily favour not panicking, because you don't know where your library is going to be used, and panicking is not a well encapsulated error communication strategy. `thiserror` or similar would be my goto here, to provide composable, structured errors on the library's public interface.

#### CLI

Hopefully you followed the advice of your libraries not panicking, because you also don't want to panic here.  When a CLI tool fails, we need to provide a human-readable error message with context (and not the kind of context you get with `panic!`, which is a stacktrace). `anyhow` is a good crate for boxed errors: An `anyhow::Error` can hold errors of any structured error type (anything that implements `Error`, which `thiserror` has a `derive` for), and also allows you to annotate error results with context, so that you can explain where in the operation things went wrong (without resorting to a stacktrace).

#### Tests (and Examples)

Panic early, panic often, panic with reckless abandon (as long as we're talking purely about test or example code -- i.e. it's not a library.  If so, see above).  Reason being -- the context you need when something goes wrong in a test is the stacktrace, which is what `panic!` (or operations that panic like `unwrap` or `expect`) gives you.  If you don't do this, you'll need to adopt the CLI strategy of manually reconstructing the context using an `anyhow::Error` and `context` calls, but this is usually redundant -- we don't care as much about the error being human readable.

File: crates/sui-sdk/src/error.rs (hunk 1 above)
Comment (amnn) [reply crates/sui-sdk/src/error.rs:21]: Thanks for clarifying.  FWIW, I wouldn't expect there to be a type that has both `JsonRpcError` and `GraphqlError` -- I expect we may want to take a second go at `sui-sdk` once the GraphQL Server has stabilised instead, so we can re-architect it to take advantage of the new API.
